flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java

```diff
@@ -57,18 +59,25 @@
 	private static final long serialVersionUID = 1L;
 
 	protected static final Logger LOG = LoggerFactory.getLogger(GenericWriteAheadSink.class);
+
+	private final String id;
 	private final CheckpointCommitter committer;
-	private transient CheckpointStreamFactory.CheckpointStateOutputStream out;
 	protected final TypeSerializer<IN> serializer;
-	private final String id;
+
+	private transient CheckpointStreamFactory.CheckpointStateOutputStream out;
 	private transient CheckpointStreamFactory checkpointStreamFactory;
 
-	private ExactlyOnceState state = new ExactlyOnceState();
+	private final TreeMap<PendingCheckpointId, PendingHandle> pendingHandles = new TreeMap<>();
```

Comment:
the `PendingCheckpointId` and `PendingHandle` can be consolidated into a single object, since they are always used in conjunction, and stored in a TreeSet.